### PR TITLE
Properly handle console-only fsc options in the VS context

### LINKED ISSUE #13549: I can make Visual Studio disappear by using "Other flags" in the properties window
If you write `--version` in the "Other flags" (motivation: I wanted to see the FSC version on each compile), and if you then switch to any source file, the whole of Visual Studio disappears.

**Repro steps**

Minimal repro zip file: [CrashVisualStudio.zip](https://github.com/dotnet/fsharp/files/9167329/CrashVisualStudio.zip)

* Create a new solution + project in VS 2022
* Go to Project Properties and set "Other flags" to `--version`
* Switch back to the main source (i.e., <kbd>Ctrl+Tab</kbd> or by clicking)
* Boom! The big disappearing act is complete: **it crashes without any warning**.

Second repro:

* Assume the above `fsproj` change is saved
* Try to open the solution of project in VS 2022
* Bring focus to any source file in that project
* Boom! VS disappears again: **it crashes without any warning**.

See video capture (sorry for the delay, I hit Ctrl-S to save, as that is often another way of crashing):

img

**Expected behavior**

Not sure if `--version` is valid here, but valid or not, Visual Studio (version 2022) should **not** disappear.

**Actual behavior**

Visual Studio crashes without a crash report. Sometimes there's an auto-restart which will hang, most of the times this restart does not happen. Sometimes child processes are killed, sometimes they remain around (like `MSBuild.exe`).

**Known workarounds**

Don't use "additional compiler options" aka "other flags".

**Related information**

I only tested this on a Windows 10 machine, I could check a Windows 11 machine for good measure. Didn't test with VS Code, only VS 2022 Community Edition.

### PR DESCRIPTION
**Pull request by psfinaki**
fixes #13549 

Changes the notion of help options to the console-only related options, which now include `--help` and `--version`.
Also added a test (best effort).

Video proof :D

[Video proof](https://user-images.githubusercontent.com/5451366/184922869-8ec969d6-dcb9-4b1a-9f48-f18406daba10.mp4)

### CODE REVIEW

File: src/Compiler/Driver/CompilerOptions.fs [hunk 1]
```diff
@@ -1983,13 +1983,13 @@ let displayVersion tcConfigB =
 let miscFlagsBoth tcConfigB =
     [
         CompilerOption("nologo", tagNone, OptionUnit(fun () -> tcConfigB.showBanner <- false), None, Some(FSComp.SR.optsNologo ()))
-        CompilerOption("version", tagNone, OptionUnit(fun () -> displayVersion tcConfigB), None, Some(FSComp.SR.optsVersion ()))
+        CompilerOption("version", tagNone, OptionConsoleOnly(fun _ -> displayVersion tcConfigB), None, Some(FSComp.SR.optsVersion ()))
```
Comment (psfinaki) [inline src/Compiler/Driver/CompilerOptions.fs:1986]: @abelbraaksma @vzarytovskii this is the actual fix, everything else is refactoring.

Comment (abelbraaksma) [reply src/Compiler/Driver/CompilerOptions.fs:1986]: Got it, tx. Does this mean using `-version` or any illegal option, will now be simply ignored? Or would it allow outputting the version info on each build to the VS console?

**Comment by abelbraaksma** [comment]
Copying here, as the ‘resolved’ review comments get hidden:

Got it, tx. Does this mean using -version or any illegal option, will now be simply ignored? Or would it allow outputting the version info on each build to the VS console?

**Comment by KevinRansom** [comment]
@psfinaki ---  hey there, there is another case, and it's a little bit wierder.

The command fsc --langversion:4.7 allows you to use ? to get the list of valid versions and then does an exit 1

image

The handler is here:
https://github.com/dotnet/fsharp/blob/main/src/Compiler/Driver/CompilerOptions.fs#L1078

And the specification is here:
https://github.com/dotnet/fsharp/blob/main/src/Compiler/Driver/CompilerOptions.fs#L1111

I think it may be a lit bit different from your CONSOLE_ONLY items, because it can also contain real options, as well as the help request.

**Comment by psfinaki** [comment]
@KevinRansom Yep, thanks for the point, I think I will address that separately in a followup. That thing is just ridiculous - the function is called "setLanguageVersion" but actually it sometimes sets the version and sometimes gets the versions. That's like against all the coding principles, so I'll better do a small refactoring and testing around that.

**Comment by psfinaki** [comment]
@abelbraaksma so normally invalid compiler options are reported. Both in console:
image
and in VS:
image

It's just that you hit a special/reserved case with this `--version` :) This is for getting info about the compiler itself, not giving it any instructions. And so I don't think it makes sense to output this info when building a project - because this info would be unrelated to the project.

So the current implementation will preserve reporting errors in VS for bad options whereas for these special readonly options VS will be just silent. Now as I'm thinking about, I might actually come with a followup to somehow point this out instead of remaining silent - but let this PR just fix the bug itself.

**Comment by vzarytovskii** [comment]
On the review with @KevinRansom, we've realized that there's an existing mechanism for proper exit from the compiler - using of exiters (with StopProcessing exceptions probably).

For example:
https://github.com/dotnet/fsharp/blob/af0015e5cac978bef2aa10264f63c886a2532bda/src/fsc/fscmain.fs#L77-L108

or:

https://github.com/dotnet/fsharp/blob/af0015e5cac978bef2aa10264f63c886a2532bda/src/Compiler/Service/service.fs#L119-L129

Wondering how easy will it be to plug into the cli options, since it is designed for cases like this and is probably the right way of dealing with it.

**Comment by vzarytovskii** [comment]
Ok, I'm going to merge it, and let's have the one with exiter next.

**Review by vzarytovskii — APPROVED** [review]

**Comment by abelbraaksma** [comment]
I finally had some time actually test this locally. Excellent work @psfinaki, solution works like a charm!

**Comment by psfinaki** [comment]
@abelbraaksma, my pleasure :)

**Pull request merged**